UnsupportedOperationException from AwsServletContext.getNamedDispatcher

*Serverless Java Container version*: `1.9.1`

*Implementations:* `Spring Boot 2`

*Framework version:* `SpringBoot 2.7.5`

*Frontend service:* `HTTP API`

*Deployment method:* `SAM`

## Scenario
Trying to get a https://hilla.dev/ Spring Boot 2 application up and running

## Expected behavior
A request to the application ends up in the deployed servlet

## Actual behavior
Every request fails with
```
Caused by: java.lang.UnsupportedOperationException: null
--
  | 2022-12-16T09:44:18.518+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsServletContext.getNamedDispatcher(AwsServletContext.java:174) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:127) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964) ~[task/:na]
  | 2022-12-16T09:44:18.518+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[task/:na]
```

This originates in `ServletForwardingController` which does
```
	protected ModelAndView handleRequestInternal(HttpServletRequest request, HttpServletResponse response)
			throws Exception {

		ServletContext servletContext = getServletContext();
		Assert.state(servletContext != null, "No ServletContext");
		RequestDispatcher rd = servletContext.getNamedDispatcher(this.servletName);
		if (rd == null) {
			throw new ServletException("No servlet with name '" + this.servletName + "' defined in web.xml");
		}
...
```

Am I holding it wrong as it seems this can never work as `getNamedDispatcher` in `AwsServletContext` always throws an exception?


## Steps to reproduce
Example project is at https://github.com/Artur-/hilla-react-lambda

## Full log output

```


2022-12-16T10:56:18.641+02:00 | START RequestId: 78cce961-568c-4604-8442-ece5862376cc Version: $LATEST
-- | --
  | 2022-12-16T10:56:18.652+02:00 | 2022-12-16 08:56:18.645 ERROR 8 --- [ main] c.a.s.p.internal.LambdaContainerHandler : Error while handling request
  | 2022-12-16T10:56:18.652+02:00 | org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.UnsupportedOperationException
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:645) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:750) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainManager$ServletExecutionFilter.doFilter(FilterChainManager.java:356) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsLambdaServletContainerHandler.doFilter(AwsLambdaServletContainerHandler.java:156) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:180) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:53) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxy(LambdaContainerHandler.java:214) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxyStream(LambdaContainerHandler.java:257) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.example.application.LambdaHandler.handleRequest(LambdaHandler.java:29) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$StreamMethodRequestHandler.handleRequest(EventHandlerLoader.java:375) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$2.call(EventHandlerLoader.java:899) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:268) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:206) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.main(AWSLambda.java:200) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | Caused by: java.lang.UnsupportedOperationException: null
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsServletContext.getNamedDispatcher(AwsServletContext.java:174) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:127) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | ... 30 common frames omitted
  | 2022-12-16T10:56:18.652+02:00 | 2022-12-16 08:56:18.645 ERROR 8 --- [ main] c.a.s.proxy.AwsProxyExceptionHandler : Called exception handler for:
  | 2022-12-16T10:56:18.652+02:00 | org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.UnsupportedOperationException
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:645) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:750) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainManager$ServletExecutionFilter.doFilter(FilterChainManager.java:356) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsLambdaServletContainerHandler.doFilter(AwsLambdaServletContainerHandler.java:156) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:180) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:53) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxy(LambdaContainerHandler.java:214) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxyStream(LambdaContainerHandler.java:257) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at com.example.application.LambdaHandler.handleRequest(LambdaHandler.java:29) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$StreamMethodRequestHandler.handleRequest(EventHandlerLoader.java:375) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$2.call(EventHandlerLoader.java:899) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:268) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:206) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.main(AWSLambda.java:200) ~[aws-lambda-java-runtime-0.2.0.jar:na]
  | 2022-12-16T10:56:18.652+02:00 | Caused by: java.lang.UnsupportedOperationException: null
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsServletContext.getNamedDispatcher(AwsServletContext.java:174) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:127) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[task/:na]
  | 2022-12-16T10:56:18.652+02:00 | ... 30 common frames omitted
  | 2022-12-16T10:56:18.652+02:00 | org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.UnsupportedOperationException
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
  | 2022-12-16T10:56:18.652+02:00 | at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainManager$ServletExecutionFilter.doFilter(FilterChainManager.java:356)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.FilterChainHolder.doFilter(FilterChainHolder.java:90)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsLambdaServletContainerHandler.doFilter(AwsLambdaServletContainerHandler.java:156)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:180)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.spring.SpringBootLambdaContainerHandler.handleRequest(SpringBootLambdaContainerHandler.java:53)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxy(LambdaContainerHandler.java:214)
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.LambdaContainerHandler.proxyStream(LambdaContainerHandler.java:257)
  | 2022-12-16T10:56:18.652+02:00 | at com.example.application.LambdaHandler.handleRequest(LambdaHandler.java:29)
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
  | 2022-12-16T10:56:18.652+02:00 | at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
  | 2022-12-16T10:56:18.652+02:00 | at java.base/java.lang.reflect.Method.invoke(Unknown Source)
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$StreamMethodRequestHandler.handleRequest(EventHandlerLoader.java:375)
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.EventHandlerLoader$2.call(EventHandlerLoader.java:899)
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:268)
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:206)
  | 2022-12-16T10:56:18.652+02:00 | at lambdainternal.AWSLambda.main(AWSLambda.java:200)
  | 2022-12-16T10:56:18.652+02:00 | Caused by: java.lang.UnsupportedOperationException
  | 2022-12-16T10:56:18.652+02:00 | at com.amazonaws.serverless.proxy.internal.servlet.AwsServletContext.getNamedDispatcher(AwsServletContext.java:174)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:127)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
  | 2022-12-16T10:56:18.652+02:00 | at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
  | 2022-12-16T10:56:18.652+02:00 | ... 30 more
  | 2022-12-16T10:56:18.653+02:00 | END RequestId: 78cce961-568c-4604-8442-ece5862376cc


```

--- a/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsProxyRequestDispatcher.java
+++ b/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsProxyRequestDispatcher.java
@@ -80,7 +80,7 @@ public class AwsProxyRequestDispatcher implements RequestDispatcher {
         }
 
         if (isNamedDispatcher) {
-            lambdaContainerHandler.doFilter((HttpServletRequest) servletRequest, (HttpServletResponse) servletResponse, getServlet((HttpServletRequest)servletRequest));
+            lambdaContainerHandler.doFilter((HttpServletRequest) servletRequest, (HttpServletResponse) servletResponse, getServlet(dispatchTo));
             return;
         }
 
@@ -148,4 +148,9 @@ public class AwsProxyRequestDispatcher implements RequestDispatcher {
     private Servlet getServlet(HttpServletRequest req) {
         return ((AwsServletContext)lambdaContainerHandler.getServletContext()).getServletForPath(req.getPathInfo());
     }
+
+    private Servlet getServlet(String servletName) throws ServletException {
+        return ((AwsServletContext)lambdaContainerHandler.getServletContext()).getServlet(servletName);
+    }
+
 }

--- a/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsServletContext.java
+++ b/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsServletContext.java
@@ -171,7 +171,7 @@ public class AwsServletContext
 
     @Override
     public RequestDispatcher getNamedDispatcher(String s) {
-        throw new UnsupportedOperationException();
+        return new AwsProxyRequestDispatcher(s, true, containerHandler);
     }
 
 

--- a/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainManager.java
+++ b/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainManager.java
@@ -132,7 +132,7 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
             chainHolder.addFilter(new FilterHolder(new ServletExecutionFilter(servletRegistration), servletContext));
         }
 
-        putFilterChainCache(type, targetPath, chainHolder);
+        putFilterChainCache(type, targetPath, servlet, chainHolder);
         // update total filter size
         if (filtersSize != registrations.size()) {
             filtersSize = registrations.size();
@@ -151,13 +151,16 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
      * initialized with the cached list of {@link FilterHolder} objects
      * @param type The dispatcher type for the incoming request
      * @param targetPath The request path - this is extracted with the <code>getPath</code> method of the request object
-     * @param servlet Servlet to put at the end of the chain (optional).
+     * @param servlet The final servlet in the filter chain (if any)
      * @return A populated FilterChainHolder
      */
     private FilterChainHolder getFilterChainCache(final DispatcherType type, final String targetPath, Servlet servlet) {
         TargetCacheKey key = new TargetCacheKey();
         key.setDispatcherType(type);
         key.setTargetPath(targetPath);
+        if (servlet != null) {
+            key.setServletName(servlet.getServletConfig().getServletName());
+        }
 
         if (!filterCache.containsKey(key)) {
             return null;
@@ -174,12 +177,16 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
      * method to retry this.
      * @param type DispatcherType from the incoming request
      * @param targetPath The target path in the API
-     * @param holder The FilterChainHolder object to save in the cache
+     * @param servlet The final servlet in the filter chain (if any)
+    * @param holder The FilterChainHolder object to save in the cache
      */
-    private void putFilterChainCache(final DispatcherType type, final String targetPath, final FilterChainHolder holder) {
+    private void putFilterChainCache(final DispatcherType type, final String targetPath, Servlet servlet, final FilterChainHolder holder) {
         TargetCacheKey key = new TargetCacheKey();
         key.setDispatcherType(type);
         key.setTargetPath(targetPath);
+        if (servlet != null) {
+            key.setServletName(servlet.getServletConfig().getServletName());
+        }
 
         // we couldn't compute the hash code because either the target path or dispatcher type were null
         if (key.hashCode() == -1) {
@@ -256,6 +263,7 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
 
         private String targetPath;
         private DispatcherType dispatcherType;
+        private String servletName;
 
 
     //-------------------------------------------------------------
@@ -295,8 +303,13 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
             }
             hashString += ":" + hashDispatcher;
 
+            if (servletName != null) {
+                hashString += ":" + servletName;
+            }
+
             return hashString.hashCode();
         }
+
 
 
         @Override
@@ -324,6 +337,11 @@ public abstract class FilterChainManager<ServletContextType extends ServletConte
         void setDispatcherType(DispatcherType dispatcherType) {
             this.dispatcherType = dispatcherType;
         }
+        public void setServletName(String servletName) {
+            this.servletName = servletName;
+        }
+
+
     }
 
     @SuppressFBWarnings("URF_UNREAD_FIELD")

--- a/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/testutils/MockServlet.java
+++ b/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/testutils/MockServlet.java
@@ -1,0 +1,23 @@
+package com.amazonaws.serverless.proxy.internal.testutils;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class MockServlet extends HttpServlet {
+
+    private int serviceCalls = 0;
+
+    @Override
+    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        super.service(req, resp);
+        serviceCalls++;
+    }
+
+    public int getServiceCalls() {
+        return serviceCalls;
+    }
+}

--- a/aws-serverless-java-container-core/src/test/java/com/amazonaws/serverless/proxy/internal/servlet/AwsFilterChainManagerTest.java
+++ b/aws-serverless-java-container-core/src/test/java/com/amazonaws/serverless/proxy/internal/servlet/AwsFilterChainManagerTest.java
@@ -2,6 +2,7 @@ package com.amazonaws.serverless.proxy.internal.servlet;
 
 import com.amazonaws.serverless.proxy.internal.testutils.AwsProxyRequestBuilder;
 import com.amazonaws.serverless.proxy.internal.testutils.MockLambdaContext;
+import com.amazonaws.serverless.proxy.internal.testutils.MockServlet;
 import com.amazonaws.services.lambda.runtime.Context;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -17,6 +18,8 @@ import java.util.concurrent.CountDownLatch;
 import static org.junit.jupiter.api.Assertions.*;
 
 public class AwsFilterChainManagerTest {
+    private static final String SERVLET1_NAME = "Servlet 1";
+    private static final String SERVLET2_NAME = "Servlet 2";
     private static final String REQUEST_CUSTOM_ATTRIBUTE_NAME = "X-Custom-Attribute";
     private static final String REQUEST_CUSTOM_ATTRIBUTE_VALUE = "CustomAttrValue";
 
@@ -36,6 +39,10 @@ public class AwsFilterChainManagerTest {
         reg2.addMappingForUrlPatterns(EnumSet.of(DispatcherType.REQUEST), true, "/second/*");
         FilterRegistration.Dynamic reg3 = servletContext.addFilter("Filter3", new MockFilter());
         reg3.addMappingForUrlPatterns(EnumSet.of(DispatcherType.REQUEST), true, "/third/fourth/*");
+        ServletRegistration.Dynamic firstServlet = servletContext.addServlet(SERVLET1_NAME, new MockServlet());
+        firstServlet.addMapping("/first/*");
+        ServletRegistration.Dynamic secondServlet = servletContext.addServlet(SERVLET2_NAME, new MockServlet());
+        secondServlet.addMapping("/second/*");
 
         chainManager = new AwsFilterChainManager((AwsServletContext) servletContext);
     }
@@ -83,6 +90,22 @@ public class AwsFilterChainManagerTest {
         FilterChainManager.TargetCacheKey secondCacheKey = new FilterChainManager.TargetCacheKey();
         secondCacheKey.setDispatcherType(DispatcherType.ASYNC);
         secondCacheKey.setTargetPath("/first/path");
+
+        assertNotEquals(cacheKey.hashCode(), secondCacheKey.hashCode());
+        assertNotEquals(cacheKey, secondCacheKey);
+    }
+
+    @Test
+    void cacheKey_compare_differentServlet() {
+        FilterChainManager.TargetCacheKey cacheKey = new FilterChainManager.TargetCacheKey();
+        cacheKey.setDispatcherType(DispatcherType.REQUEST);
+        cacheKey.setTargetPath("/first/path");
+        cacheKey.setServletName("Dispatcher servlet");
+
+        FilterChainManager.TargetCacheKey secondCacheKey = new FilterChainManager.TargetCacheKey();
+        secondCacheKey.setDispatcherType(DispatcherType.REQUEST);
+        secondCacheKey.setTargetPath("/first/path");
+        cacheKey.setServletName("Real servlet");
 
         assertNotEquals(cacheKey.hashCode(), secondCacheKey.hashCode());
         assertNotEquals(cacheKey, secondCacheKey);
@@ -154,7 +177,7 @@ public class AwsFilterChainManagerTest {
     }
 
     @Test
-    void filerChain_executeMultipleFilters_expectRunEachTime() {
+    void filterChain_executeMultipleFilters_expectRunEachTime() {
         AwsProxyHttpServletRequest req = new AwsProxyHttpServletRequest(
                 new AwsProxyRequestBuilder("/first/second", "GET").build(), lambdaContext, null
         );
@@ -202,6 +225,34 @@ public class AwsFilterChainManagerTest {
 
         assertTrue(req2.getAttribute(REQUEST_CUSTOM_ATTRIBUTE_NAME) != null);
         assertEquals(REQUEST_CUSTOM_ATTRIBUTE_VALUE, req2.getAttribute(REQUEST_CUSTOM_ATTRIBUTE_NAME));
+    }
+
+    @Test
+    void filterChain_multipleServlets_callsCorrectServlet() throws IOException, ServletException {
+        MockServlet servlet1 = (MockServlet) servletContext.getServlet(SERVLET1_NAME);
+        ServletConfig servlet1Config = ((AwsServletRegistration) servletContext.getServletRegistration(SERVLET1_NAME)).getServletConfig();
+        servlet1.init(servlet1Config);
+
+        MockServlet servlet2 = (MockServlet) servletContext.getServlet(SERVLET2_NAME);
+        ServletConfig servlet2Config = ((AwsServletRegistration) servletContext.getServletRegistration(SERVLET2_NAME)).getServletConfig();
+        servlet2.init(servlet2Config);
+
+        AwsProxyHttpServletRequest req = new AwsProxyHttpServletRequest(
+                new AwsProxyRequestBuilder("/", "GET").build(), lambdaContext, null
+        );
+        AwsHttpServletResponse resp = new AwsHttpServletResponse(req, new CountDownLatch(1));
+
+        FilterChainHolder servlet1filterChain = chainManager.getFilterChain(req, servlet1);
+        servlet1filterChain.doFilter(req, resp);
+        
+        assertEquals(1, servlet1.getServiceCalls());
+        assertEquals(0, servlet2.getServiceCalls());
+
+        FilterChainHolder servlet2filterChain = chainManager.getFilterChain(req, servlet2);
+        servlet2filterChain.doFilter(req, resp);
+
+        assertEquals(1, servlet1.getServiceCalls());
+        assertEquals(1, servlet2.getServiceCalls());
     }
 
     @Test

--- a/aws-serverless-java-container-core/src/test/java/com/amazonaws/serverless/proxy/internal/servlet/AwsServletContextTest.java
+++ b/aws-serverless-java-container-core/src/test/java/com/amazonaws/serverless/proxy/internal/servlet/AwsServletContextTest.java
@@ -13,12 +13,7 @@ import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 import java.io.File;
-import java.io.FileNotFoundException;
 import java.io.IOException;
-import java.io.PrintWriter;
-import java.io.UnsupportedEncodingException;
-import java.nio.file.Files;
-import java.nio.file.Paths;
 import java.util.concurrent.CountDownLatch;
 
 import static org.junit.jupiter.api.Assertions.*;
@@ -190,12 +185,7 @@ public class AwsServletContextTest {
         } catch (UnsupportedOperationException e) {
             exCount++;
         }
-        try {
-            STATIC_CTX.getNamedDispatcher("1");
-        } catch (UnsupportedOperationException e) {
-            exCount++;
-        }
-        assertEquals(2, exCount);
+        assertEquals(1, exCount);
 
         assertNull(STATIC_CTX.getServletRegistration("1"));
     }
@@ -230,6 +220,12 @@ public class AwsServletContextTest {
         assertNotNull(ctx.getServlet("srv1"));
         assertNotNull(ctx.getServletRegistration("srv1"));
         assertEquals("", ((TestServlet)ctx.getServlet("srv1")).getId());
+    }
+
+    @Test
+    void getNamedDispatcher_returnsDispatcher() {
+        AwsServletContext ctx = new AwsServletContext(null);
+        assertNotNull(ctx.getNamedDispatcher("/hello"));
     }
 
     public static class TestServlet implements Servlet {

